File: Abracadabra/RoundObfusHelper.js
/*
 * Copyright (C) 2025 SheepChef (a.k.a. Haruka Hokuto)
 *
 * 这是一个自由软件。
 * 在遵守AIPL-1.1许可证的前提下，
 * 你可以自由复制，修改，分发，使用它。
 *
 * 查阅 Academic Innovation Protection License(AIPL) 来了解更多 .
 * 本作品应随附一份完整的 AIPL-1.1 许可证全文。
 *
 */

import CryptoJS from "crypto-js";
import { wordArrayToUint8Array } from "./Misc.js";
import { CallbackObj } from "./CoreHandler.js";

export class RoundObfus {
  constructor(key, callback = null) {
    this.RoundFlip = 0; //标志现在到哪了
    this.RoundControl = new Uint8Array(32); //一个数组，用密钥哈希来控制轮转的行为
    this.LETTERS_ROUND_1 =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    this.LETTERS_ROUND_2 =
      "FbPoDRStyJKAUcdahfVXlqwnOGpHZejzvmrBCigQILxkYMuWTEsN"; //手动随机打乱的乱序轮
    this.LETTERS_ROUND_3 =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    this.NUMBERSYMBOL_ROUND_1 = "1234567890+/=";
    this.NUMBERSYMBOL_ROUND_2 = "5=0764+389/12"; //手动随机打乱的乱序轮
    this.NUMBERSYMBOL_ROUND_3 = "1234567890+/=";

    this.Normal_Characters =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/=1234567890"; //表内有映射的所有字符组成的字符串
    this.LETTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

    this.BIG_LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    this.NUMBERS = "1234567890";
    this.SYMBOLS = "+/=";
    this.NUMBERSYMBOL = "0123456789+/=";

    this.NULL_STR = "孎"; //默认忽略的占位字符，一个生僻字。

    this.callback = callback;

    //初始化转轮操作的数组
    let KeyHash = CryptoJS.SHA256(key);
    let HashArray = wordArrayToUint8Array(KeyHash);

    this.RoundControl = HashArray;
  }

  _rotateString(str, n) {
    // 向右轮转指定位数
    return str.slice(n) + str.slice(0, n);
  }

  _LrotateString(str, n) {
    // 向左轮转指定位数
    return str.slice(str.length - n) + str.slice(0, str.length - n);
  }
  RoundKeyMatch(keyIn) {
    //查询轮换密钥的键值
    let idx1, idx2;
    let idx1_1, idx2_1;
    let idx1_2, idx2_2;

    idx1 = this.LETTERS.indexOf(keyIn);
    idx2 = this.NUMBERSYMBOL.indexOf(keyIn);

    idx1_1 = this.LETTERS.indexOf(this.LETTERS_ROUND_1[idx1]);
    idx2_1 = this.NUMBERSYMBOL.indexOf(this.NUMBERSYMBOL_ROUND_1[idx2]);

    idx1_2 = this.LETTERS.indexOf(this.LETTERS_ROUND_2[idx1_1]);
    idx2_2 = this.NUMBERSYMBOL.indexOf(this.NUMBERSYMBOL_ROUND_2[idx2_1]);

    if (idx1 != -1) {
      //判断给定字符的类型
      return this.LETTERS_ROUND_3[idx1_2];
    } else if (idx2 != -1) {
      return this.NUMBERSYMBOL_ROUND_3[idx2_2];
    }
    return this.NULL_STR;
  }

  DRoundKeyMatch(keyIn) {
    //查询轮换密钥的键值
    let idx1, idx2;
    let idx1_1, idx2_1;
    let idx1_2, idx2_2;

    idx1 = this.LETTERS_ROUND_3.indexOf(keyIn);
    idx2 = this.NUMBERSYMBOL_ROUND_3.indexOf(keyIn);

    idx1_1 = this.LETTERS_ROUND_2.indexOf(this.LETTERS[idx1]);
    idx2_1 = this.NUMBERSYMBOL_ROUND_2.indexOf(this.NUMBERSYMBOL[idx2]);

    idx1_2 = this.LETTERS_ROUND_1.indexOf(this.LETTERS[idx1_1]);
    idx2_2 = this.NUMBERSYMBOL_ROUND_1.indexOf(this.NUMBERSYMBOL[idx2_1]);

    if (idx1 != -1) {
      //判断给定字符的类型
      return this.LETTERS[idx1_2];
    } else if (idx2 != -1) {
      return this.NUMBERSYMBOL[idx2_2];
    }
    return this.NULL_STR;
  }

  RoundKey() {
    let ControlNum = 0;
    if (this.RoundFlip == 32) {
      this.RoundFlip = 0;
    }
    ControlNum = this.RoundControl[this.RoundFlip] % 10; //哈希字节对十取余即操作数
    if (ControlNum == 0) {
      //等于零就赋值为10
      ControlNum = 10;
    }

    if (ControlNum % 2 == 0) {
      //操作数是偶数
      this.LETTERS_ROUND_1 = this._rotateString(this.LETTERS_ROUND_1, 6); //将第一个密钥轮向右轮6位
      this.NUMBERSYMBOL_ROUND_1 = this._rotateString(
        this.NUMBERSYMBOL_ROUND_1,
        6
      );

      this.LETTERS_ROUND_2 = this._LrotateString(
        this.LETTERS_ROUND_2,
        ControlNum
      ); //将第二个密钥轮向左轮ControlNum*2位
      this.NUMBERSYMBOL_ROUND_2 = this._LrotateString(
        this.NUMBERSYMBOL_ROUND_2,
        ControlNum
      );

      this.LETTERS_ROUND_3 = this._rotateString(
        this.LETTERS_ROUND_3,
        ControlNum / 2 + 1
      ); //将第三个密钥轮向右轮ControlNum/2+1位
      this.NUMBERSYMBOL_ROUND_3 = this._rotateString(
        this.NUMBERSYMBOL_ROUND_3,
        ControlNum / 2 + 1
      );
    } else {
      //操作数是奇数
      this.LETTERS_ROUND_1 = this._LrotateString(this.LETTERS_ROUND_1, 3); //将第一个密钥轮向左轮3位
      this.NUMBERSYMBOL_ROUND_1 = this._LrotateString(
        this.NUMBERSYMBOL_ROUND_1,
        3
      );

      this.LETTERS_ROUND_2 = this._rotateString(
        this.LETTERS_ROUND_2,
        ControlNum
      ); //将第二个密钥轮向右轮ControlNum位
      this.NUMBERSYMBOL_ROUND_2 = this._rotateString(
        this.NUMBERSYMBOL_ROUND_2,
        ControlNum
      );

      this.LETTERS_ROUND_3 = this._LrotateString(
        this.LETTERS_ROUND_3,
        (ControlNum + 7) / 2
      ); //将第三个密钥轮向左轮(ControlNum+5)/2位
      this.NUMBERSYMBOL_ROUND_3 = this._LrotateString(
        this.NUMBERSYMBOL_ROUND_3,
        (ControlNum + 7) / 2
      );
    }
    this.RoundFlip++;
    try {
      if (this.callback != null)
        this.callback(
          new CallbackObj("ROUNDS", [
            this.LETTERS_ROUND_1,
            this.LETTERS_ROUND_2,
            this.LETTERS_ROUND_3,
            this.NUMBERSYMBOL_ROUND_1,
            this.NUMBERSYMBOL_ROUND_2,
            this.NUMBERSYMBOL_ROUND_3,
          ])
        );
    } catch (err) {}
  }
}

export class RoundObfusOLD {
  constructor(key) {
    this.RoundFlip = 0; //标志现在到哪了
    this.RoundControl = new Uint8Array(32); //一个数组，用密钥哈希来控制轮转的行为
    this.LETTERS_ROUND_1 =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    this.LETTERS_ROUND_2 =
      "FbPoDRStyJKAUcdahfVXlqwnOGpHZejzvmrBCigQILxkYMuWTEsN"; //手动随机打乱的乱序轮
    this.LETTERS_ROUND_3 =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
    this.NUMBERSYMBOL_ROUND_1 = "1234567890+=_-/?.>,<|`~!@#$%^&*(){}[];:";
    this.NUMBERSYMBOL_ROUND_2 = "~3{8}_-$[6(2^&#5|1*%0,<9:`+@7/?.>4=];!)"; //手动随机打乱的乱序轮
    this.NUMBERSYMBOL_ROUND_3 = "1234567890+=_-/?.>,<|`~!@#$%^&*(){}[];:";

    this.Normal_Characters =
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+=_-/?.>,<|`~!@#$%^&*(){}[];:1234567890"; //表内有映射的所有字符组成的字符串
    this.LETTERS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

    this.BIG_LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    this.NUMBERS = "1234567890";
    this.SYMBOLS = "+=_-/?.>,<|`~!@#$%^&*(){}[];:";
    this.NUMBERSYMBOL = "1234567890+=_-/?.>,<|`~!@#$%^&*(){}[];:";

    this.NULL_STR = "孎"; //默认忽略的占位字符，一个生僻字。
    //初始化转轮操作的数组
    let KeyHash = CryptoJS.SHA256(key);
    let HashArray = wordArrayToUint8Array(KeyHash);

    this.RoundControl = HashArray;
  }

  _rotateString(str, n) {
    // 向右轮转指定位数
    return str.slice(n) + str.slice(0, n);
  }

  _LrotateString(str, n) {
    // 向左轮转指定位数
    return str.slice(str.length - n) + str.slice(0, str.length - n);
  }
  RoundKeyMatch(keyIn) {
    //查询轮换密钥的键值
    let idx1, idx2;
    let idx1_1, idx2_1;
    let idx1_2, idx2_2;

    idx1 = this.LETTERS.indexOf(keyIn);
    idx2 = this.NUMBERSYMBOL.indexOf(keyIn);

    idx1_1 = this.LETTERS.indexOf(this.LETTERS_ROUND_1[idx1]);
    idx2_1 = this.NUMBERSYMBOL.indexOf(this.NUMBERSYMBOL_ROUND_1[idx2]);

    idx1_2 = this.LETTERS.indexOf(this.LETTERS_ROUND_2[idx1_1]);
    idx2_2 = this.NUMBERSYMBOL.indexOf(this.NUMBERSYMBOL_ROUND_2[idx2_1]);

    if (idx1 != -1) {
      //判断给定字符的类型
      return this.LETTERS_ROUND_3[idx1_2];
    } else if (idx2 != -1) {
      return this.NUMBERSYMBOL_ROUND_3[idx2_2];
    }
    return this.NULL_STR;
  }

  DRoundKeyMatch(keyIn) {
    //查询轮换密钥的键值
    let idx1, idx2;
    let idx1_1, idx2_1;
    let idx1_2, idx2_2;

    idx1 = this.LETTERS_ROUND_3.indexOf(keyIn);
    idx2 = this.NUMBERSYMBOL_ROUND_3.indexOf(keyIn);

    idx1_1 = this.LETTERS_ROUND_2.indexOf(this.LETTERS[idx1]);
    idx2_1 = this.NUMBERSYMBOL_ROUND_2.indexOf(this.NUMBERSYMBOL[idx2]);

    idx1_2 = this.LETTERS_ROUND_1.indexOf(this.LETTERS[idx1_1]);
    idx2_2 = this.NUMBERSYMBOL_ROUND_1.indexOf(this.NUMBERSYMBOL[idx2_1]);

    if (idx1 != -1) {
      //判断给定字符的类型
      return this.LETTERS[idx1_2];
    } else if (idx2 != -1) {
      return this.NUMBERSYMBOL[idx2_2];
    }
    return this.NULL_STR;
  }

  RoundKey() {
    let ControlNum = 0;
    if (this.RoundFlip == 32) {
      this.RoundFlip = 0;
    }
    ControlNum = this.RoundControl[this.RoundFlip] % 10; //哈希字节对十取余即操作数
    if (ControlNum == 0) {
      //等于零就赋值为10
      ControlNum = 10;
    }

    if (ControlNum % 2 == 0) {
      //操作数是偶数
      this.LETTERS_ROUND_1 = this._rotateString(this.LETTERS_ROUND_1, 6); //将第一个密钥轮向右轮6位
      this.NUMBERSYMBOL_ROUND_1 = this._rotateString(
        this.NUMBERSYMBOL_ROUND_1,
        6
      );

      this.LETTERS_ROUND_2 = this._LrotateString(
        this.LETTERS_ROUND_2,
        ControlNum * 2
      ); //将第二个密钥轮向左轮ControlNum*2位
      this.NUMBERSYMBOL_ROUND_2 = this._LrotateString(
        this.NUMBERSYMBOL_ROUND_2,
        ControlNum * 2
      );

      this.LETTERS_ROUND_3 = this._rotateString(
        this.LETTERS_ROUND_3,
        ControlNum / 2 + 1
      ); //将第三个密钥轮向右轮ControlNum/2+1位
      this.NUMBERSYMBOL_ROUND_3 = this._rotateString(
        this.NUMBERSYMBOL_ROUND_3,
        ControlNum / 2 + 1
      );
    } else {
      //操作数是奇数
      this.LETTERS_ROUND_1 = this._LrotateString(this.LETTERS_ROUND_1, 3); //将第一个密钥轮向左轮3位
      this.NUMBERSYMBOL_ROUND_1 = this._LrotateString(
        this.NUMBERSYMBOL_ROUND_1,
        3
      );

      this.LETTERS_ROUND_2 = this._rotateString(
        this.LETTERS_ROUND_2,
        ControlNum
      ); //将第二个密钥轮向右轮ControlNum位
      this.NUMBERSYMBOL_ROUND_2 = this._rotateString(
        this.NUMBERSYMBOL_ROUND_2,
        ControlNum
      );

      this.LETTERS_ROUND_3 = this._LrotateString(
        this.LETTERS_ROUND_3,
        (ControlNum + 7) / 2
      ); //将第三个密钥轮向左轮(ControlNum+5)/2位
      this.NUMBERSYMBOL_ROUND_3 = this._LrotateString(
        this.NUMBERSYMBOL_ROUND_3,
        (ControlNum + 7) / 2
      );
    }
    this.RoundFlip++;
  }
}

File: src/RoundObfusHelper.cpp
/*
 * Copyright (C) 2025 SheepChef (a.k.a. Haruka Hokuto)
 *
 * 这是一个自由软件。
 * 在遵守AIPL-1.1许可证的前提下，
 * 你可以自由复制，修改，分发，使用它。
 *
 * 查阅 Academic Innovation Protection License(AIPL) 来了解更多 .
 * 本作品应随附一份完整的 AIPL-1.1 许可证全文。
 *
 */

#include <array>
#include <cstdint>
#include <functional>
#include <string>
#include <vector>

#include "Misc.cpp"

namespace moyue::round {

namespace {
std::vector<std::uint8_t> HashKeyToBytes(const std::string& key) {
  // TODO: Replace with SHA256 hashing of the key (CryptoJS.SHA256 equivalent).
  std::vector<std::uint8_t> bytes(32, 0);
  for (size_t i = 0; i < key.size(); ++i) {
    bytes[i % bytes.size()] ^= static_cast<std::uint8_t>(key[i]);
  }
  return bytes;
}

std::string RotateString(const std::string& input, size_t count) {
  if (input.empty()) {
    return input;
  }
  size_t shift = count % input.size();
  return input.substr(shift) + input.substr(0, shift);
}

std::string RotateStringLeft(const std::string& input, size_t count) {
  if (input.empty()) {
    return input;
  }
  size_t shift = count % input.size();
  return input.substr(input.size() - shift) + input.substr(0, input.size() - shift);
}
}

using RoundCallback = std::function<void(const std::string&)>;

class RoundObfus {
 public:
  explicit RoundObfus(const std::string& key,
                      const RoundCallback& callback = nullptr)
      : roundFlip_(0),
        lettersRound1_("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        lettersRound2_(
            "FbPoDRStyJKAUcdahfVXlqwnOGpHZejzvmrBCigQILxkYMuWTEsN"),
        lettersRound3_("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        numberSymbolRound1_("1234567890+/="),
        numberSymbolRound2_("5=0764+389/12"),
        numberSymbolRound3_("1234567890+/="),
        normalCharacters_(
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+/=1234567890"),
        letters_("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        bigLetters_("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        numbers_("1234567890"),
        symbols_("+/="),
        numberSymbol_("0123456789+/="),
        nullStr_("孎"),
        callback_(callback) {
    //初始化转轮操作的数组
    auto hashArray = HashKeyToBytes(key);
    roundControl_.assign(hashArray.begin(), hashArray.end());
  }

  std::string RoundKeyMatch(const std::string& keyIn) {
    //查询轮换密钥的键值
    int idx1 = static_cast<int>(letters_.find(keyIn));
    int idx2 = static_cast<int>(numberSymbol_.find(keyIn));

    if (idx1 != -1) {
      int idx1_1 = static_cast<int>(letters_.find(lettersRound1_.at(idx1)));
      int idx1_2 = static_cast<int>(letters_.find(lettersRound2_.at(idx1_1)));
      return std::string(1, lettersRound3_.at(idx1_2));
    }
    if (idx2 != -1) {
      int idx2_1 = static_cast<int>(numberSymbol_.find(numberSymbolRound1_.at(idx2)));
      int idx2_2 = static_cast<int>(numberSymbol_.find(numberSymbolRound2_.at(idx2_1)));
      return std::string(1, numberSymbolRound3_.at(idx2_2));
    }
    return nullStr_;
  }

  std::string DRoundKeyMatch(const std::string& keyIn) {
    //查询轮换密钥的键值
    int idx1 = static_cast<int>(lettersRound3_.find(keyIn));
    int idx2 = static_cast<int>(numberSymbolRound3_.find(keyIn));

    if (idx1 != -1) {
      int idx1_1 = static_cast<int>(lettersRound2_.find(letters_.at(idx1)));
      int idx1_2 = static_cast<int>(lettersRound1_.find(letters_.at(idx1_1)));
      return std::string(1, letters_.at(idx1_2));
    }
    if (idx2 != -1) {
      int idx2_1 = static_cast<int>(numberSymbolRound2_.find(numberSymbol_.at(idx2)));
      int idx2_2 = static_cast<int>(numberSymbolRound1_.find(numberSymbol_.at(idx2_1)));
      return std::string(1, numberSymbol_.at(idx2_2));
    }
    return nullStr_;
  }

  void RoundKey() {
    int controlNum = 0;
    if (roundFlip_ == 32) {
      roundFlip_ = 0;
    }
    controlNum = roundControl_[roundFlip_] % 10; //哈希字节对十取余即操作数
    if (controlNum == 0) {
      //等于零就赋值为10
      controlNum = 10;
    }

    if (controlNum % 2 == 0) {
      //操作数是偶数
      lettersRound1_ = RotateString(lettersRound1_, 6); //将第一个密钥轮向右轮6位
      numberSymbolRound1_ = RotateString(numberSymbolRound1_, 6);

      lettersRound2_ = RotateStringLeft(lettersRound2_, controlNum);
      numberSymbolRound2_ = RotateStringLeft(numberSymbolRound2_, controlNum);

      lettersRound3_ = RotateString(lettersRound3_, controlNum / 2 + 1);
      numberSymbolRound3_ = RotateString(numberSymbolRound3_, controlNum / 2 + 1);
    } else {
      //操作数是奇数
      lettersRound1_ = RotateStringLeft(lettersRound1_, 3);
      numberSymbolRound1_ = RotateStringLeft(numberSymbolRound1_, 3);

      lettersRound2_ = RotateString(lettersRound2_, controlNum);
      numberSymbolRound2_ = RotateString(numberSymbolRound2_, controlNum);

      lettersRound3_ = RotateStringLeft(lettersRound3_, (controlNum + 7) / 2);
      numberSymbolRound3_ = RotateStringLeft(numberSymbolRound3_, (controlNum + 7) / 2);
    }
    roundFlip_++;

    if (callback_) {
      callback_(lettersRound1_ + lettersRound2_ + lettersRound3_ +
                numberSymbolRound1_ + numberSymbolRound2_ + numberSymbolRound3_);
    }
  }

 private:
  int roundFlip_;
  std::vector<std::uint8_t> roundControl_;
  std::string lettersRound1_;
  std::string lettersRound2_;
  std::string lettersRound3_;
  std::string numberSymbolRound1_;
  std::string numberSymbolRound2_;
  std::string numberSymbolRound3_;
  std::string normalCharacters_;
  std::string letters_;
  std::string bigLetters_;
  std::string numbers_;
  std::string symbols_;
  std::string numberSymbol_;
  std::string nullStr_;
  RoundCallback callback_;
};

class RoundObfusOld {
 public:
  explicit RoundObfusOld(const std::string& key)
      : roundFlip_(0),
        lettersRound1_("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        lettersRound2_(
            "FbPoDRStyJKAUcdahfVXlqwnOGpHZejzvmrBCigQILxkYMuWTEsN"),
        lettersRound3_("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        numberSymbolRound1_("1234567890+=_-/?.>,<|`~!@#$%^&*(){}[];:"),
        numberSymbolRound2_(
            "~3{8}_-$[6(2^&#5|1*%0,<9:`+@7/?.>4=];!)"),
        numberSymbolRound3_("1234567890+=_-/?.>,<|`~!@#$%^&*(){}[];:"),
        normalCharacters_(
            "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ+=_-/?.>,<|`~!@#$%^&*(){}[];:1234567890"),
        letters_("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        bigLetters_("ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
        numbers_("1234567890"),
        symbols_("+=_-/?.>,<|`~!@#$%^&*(){}[];:"),
        numberSymbol_("1234567890+=_-/?.>,<|`~!@#$%^&*(){}[];:"),
        nullStr_("孎") {
    auto hashArray = HashKeyToBytes(key);
    roundControl_.assign(hashArray.begin(), hashArray.end());
  }

  std::string RoundKeyMatch(const std::string& keyIn) {
    int idx1 = static_cast<int>(letters_.find(keyIn));
    int idx2 = static_cast<int>(numberSymbol_.find(keyIn));

    if (idx1 != -1) {
      int idx1_1 = static_cast<int>(letters_.find(lettersRound1_.at(idx1)));
      int idx1_2 = static_cast<int>(letters_.find(lettersRound2_.at(idx1_1)));
      return std::string(1, lettersRound3_.at(idx1_2));
    }
    if (idx2 != -1) {
      int idx2_1 = static_cast<int>(numberSymbol_.find(numberSymbolRound1_.at(idx2)));
      int idx2_2 = static_cast<int>(numberSymbol_.find(numberSymbolRound2_.at(idx2_1)));
      return std::string(1, numberSymbolRound3_.at(idx2_2));
    }
    return nullStr_;
  }

  std::string DRoundKeyMatch(const std::string& keyIn) {
    int idx1 = static_cast<int>(lettersRound3_.find(keyIn));
    int idx2 = static_cast<int>(numberSymbolRound3_.find(keyIn));

    if (idx1 != -1) {
      int idx1_1 = static_cast<int>(lettersRound2_.find(letters_.at(idx1)));
      int idx1_2 = static_cast<int>(lettersRound1_.find(letters_.at(idx1_1)));
      return std::string(1, letters_.at(idx1_2));
    }
    if (idx2 != -1) {
      int idx2_1 = static_cast<int>(numberSymbolRound2_.find(numberSymbol_.at(idx2)));
      int idx2_2 = static_cast<int>(numberSymbolRound1_.find(numberSymbol_.at(idx2_1)));
      return std::string(1, numberSymbol_.at(idx2_2));
    }
    return nullStr_;
  }

  void RoundKey() {
    int controlNum = 0;
    if (roundFlip_ == 32) {
      roundFlip_ = 0;
    }
    controlNum = roundControl_[roundFlip_] % 10; //哈希字节对十取余即操作数
    if (controlNum == 0) {
      //等于零就赋值为10
      controlNum = 10;
    }

    if (controlNum % 2 == 0) {
      //操作数是偶数
      lettersRound1_ = RotateString(lettersRound1_, 6);
      numberSymbolRound1_ = RotateString(numberSymbolRound1_, 6);

      lettersRound2_ = RotateStringLeft(lettersRound2_, controlNum * 2);
      numberSymbolRound2_ = RotateStringLeft(numberSymbolRound2_, controlNum * 2);

      lettersRound3_ = RotateString(lettersRound3_, controlNum / 2 + 1);
      numberSymbolRound3_ = RotateString(numberSymbolRound3_, controlNum / 2 + 1);
    } else {
      //操作数是奇数
      lettersRound1_ = RotateStringLeft(lettersRound1_, 3);
      numberSymbolRound1_ = RotateStringLeft(numberSymbolRound1_, 3);

      lettersRound2_ = RotateString(lettersRound2_, controlNum);
      numberSymbolRound2_ = RotateString(numberSymbolRound2_, controlNum);

      lettersRound3_ = RotateStringLeft(lettersRound3_, (controlNum + 7) / 2);
      numberSymbolRound3_ = RotateStringLeft(numberSymbolRound3_, (controlNum + 7) / 2);
    }
    roundFlip_++;
  }

 private:
  int roundFlip_;
  std::vector<std::uint8_t> roundControl_;
  std::string lettersRound1_;
  std::string lettersRound2_;
  std::string lettersRound3_;
  std::string numberSymbolRound1_;
  std::string numberSymbolRound2_;
  std::string numberSymbolRound3_;
  std::string normalCharacters_;
  std::string letters_;
  std::string bigLetters_;
  std::string numbers_;
  std::string symbols_;
  std::string numberSymbol_;
  std::string nullStr_;
};

}  // namespace moyue::round
